Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -1474,6 +1474,13 @@ void FrameLoader::loadURL(FrameLoadRequest&& frameLoadRequest, const String& ref
     if (m_inStopAllLoaders || m_inClearProvisionalLoadForPolicyCheck)
         return;
 
+#if PLATFORM(WPE) || PLATFORM(GTK)
+    if ((m_policyDocumentLoader && m_policyDocumentLoader->isRequestFromClientOrUserInput()) || (m_provisionalDocumentLoader && m_provisionalDocumentLoader->isRequestFromClientOrUserInput())) {
```

Comment:
So last time you talked to me about this, I misunderstood the problem. Seeing this again now, I definitely agree that the web content should not be able to cancel an API request. The API request should always win out (and on all platforms).

What I don't understand is why this particular check works. It looks like you cancel all URL loads if `isRequestFromClientOrUserInput()` returns true, which seems very broad. So I wonder: how is it this doesn't break all API requests?

Reply:
API requests go with different call stack: 
```
1 0x739fe2427b00 WebCore::FrameLoader::loadWithDocumentLoader(WebCore::DocumentLoader*, WebCore::FrameLoadType, WTF::RefPtr<WebCore::FormState, WTF::RawPtrTraitsWebCore::FormState, WTF::DefaultRefDerefTraitsWebCore::FormState >&&, WebCore::AllowNavigationToInvalidURL, WTF::CompletionHandler<void ()>&&)
2 0x739fe242bc8e WebCore::FrameLoader::load(WebCore::DocumentLoader&, WebCore::SecurityOrigin const*)
3 0x739fe242c7ba WebCore::FrameLoader::load(WebCore::FrameLoadRequest&&)
4 0x739fdefc9e08 WebKit::WebPage::loadRequest(WebKit::LoadParameters&&)
5 0x739fde9a295b void IPC::handleMessage<Messages::WebPage::LoadRequest, WebKit::WebPage, WebKit::WebPage, void (WebKit::LoadParameters&&)>(IPC::Connection&, IPC::Decoder&, WebKit::WebPage*, void (WebKit::WebPage::)(WebKit::LoadParameters&&)) [clone .isra.0]
6 0x739fde9ae97e WebKit::WebPage::didReceiveWebPageMessage(IPC::Connection&, IPC::Decoder&)
7 0x739fdec17b7a IPC::MessageReceiverMap::dispatchMessage(IPC::Connection&, IPC::Decoder&)
8 0x739fdeec859a WebKit::WebProcess::didReceiveMessage(IPC::Connection&, IPC::Decoder&)
9 0x739fdec1365d IPC::Connection::dispatchMessage(WTF::UniqueRefIPC::Decoder)
10 0x739fdec14ada IPC::Connection::dispatchOneIncomingMessage()
11 0x739fe06f698e WTF::RunLoop::performWork()
12 0x739fe07713bd WTF::RunLoop::RunLoop()::{lambda(void)#1}::_FUN(void*)
13 0x739fe07721d3 WTF::RunLoop::{lambda(_GSource*, int ()(void), void*)#1}::_FUN(_GSource*, int ()(void), void*)
14 0x739fe51993df g_main_context_dispatch
15 0x739fe51f42f8 ???
16 0x739fe5198c4f g_main_loop_run
17 0x739fe0772380 WTF::RunLoop::run()
18 0x739fdf00b08e WebKit::WebProcessMain(int, char**)
19 0x739fddc1ea90 __libc_start_call_main
20 0x739fddc1eb49 __libc_start_main
21 0x5d6342d52345 _start
```

Fragment navigation callstack:
```
1 0x7a63d74289f8 WebCore::FrameLoader::loadURL(WebCore::FrameLoadRequest&&, WTF::String const&, WebCore::FrameLoadType, WebCore::Event*, WTF::RefPtr<WebCore::FormState, WTF::RawPtrTraitsWebCore::FormState, WTF::DefaultRefDerefTraitsWebCore::FormState >&&, std::optionalWebCore::PrivateClickMeasurement&&, WTF::CompletionHandler<void ()>&&)
2 0x7a63d742acf9 WebCore::FrameLoader::loadFrameRequest(WebCore::FrameLoadRequest&&, WebCore::Event*, WTF::RefPtr<WebCore::FormState, WTF::RawPtrTraitsWebCore::FormState, WTF::DefaultRefDerefTraitsWebCore::FormState >&&, std::optionalWebCore::PrivateClickMeasurement&&)
3 0x7a63d742af45 WebCore::FrameLoader::changeLocation(WebCore::FrameLoadRequest&&, WebCore::Event*, std::optionalWebCore::PrivateClickMeasurement&&)
4 0x7a63d744fe84 WebCore::NavigationScheduler::scheduleLocationChange(WebCore::Document&, WebCore::SecurityOrigin&, WTF::URL const&, WTF::String const&, WebCore::LockHistory, WebCore::LockBackForwardList, WTF::CompletionHandler<void (WebCore::ScheduleLocationChangeResult)>&&)
5 0x7a63d7594d6f WebCore::LocalDOMWindow::setLocation(WebCore::LocalDOMWindow&, WTF::URL const&, WebCore::SetLocationLocking)
6 0x7a63d75a5208 WebCore::Location::setLocation(WebCore::LocalDOMWindow&, WebCore::LocalDOMWindow&, WTF::String const&)
7 0x7a63d75a534e WebCore::Location::setHref(WebCore::LocalDOMWindow&, WebCore::LocalDOMWindow&, WTF::String const&)
8 0x7a63d5e8fcb8 WebCore::setJSLocation_href(JSC::JSGlobalObject*, long, long, JSC::PropertyName)
9 0x7a6380d142db ???
```